### PR TITLE
setup login/logout command

### PR DESCRIPTION
**Pull request by fahedouch**
Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

### CODE REVIEW

File: login.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+type loginOptions struct {
+	serverAddress string
+	username      string
+	password      string
+	passwordStdin bool
+}
+
+func NewLoginCommand() *cli.Command {
```
Comment (AkihiroSuda) [inline login.go:35]: Was this ported from https://github.com/genuinetools/img/blob/v0.5.11/login.go ?
If so, please clarify the origin in the code comment.

File: main.go [hunk 2]
```diff
@@ -148,6 +148,8 @@ func newApp() *cli.App {
 		systemCommand,
 		// Internal
 		internalCommand,
+		// login
+		NewLoginCommand(),
```
Comment (AkihiroSuda) [inline main.go:152]: coding style seems inconsistent with other commands

Comment (fahedouch) [reply main.go:152]: @AkihiroSuda  I have to set a function here. The problem is with setting a function or a coding formatting ?

Comment (AkihiroSuda) [reply main.go:152]: Why does it need to be a function?

Comment (fahedouch) [reply main.go]: @AkihiroSuda  fixed

File: login.go (hunk 1 above)
Comment (fahedouch) [reply login.go]: @AkihiroSuda  fixed

File: login.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	"os"
+	"strings"
+
+	"github.com/docker/cli/cli/command"
+	dockercliconfig "github.com/docker/cli/cli/config"
+	clitypes "github.com/docker/cli/cli/config/types"
+	dockercliconfigtypes "github.com/docker/cli/cli/config/types"
+	"github.com/docker/docker/api/types"
+	registrytypes "github.com/docker/docker/api/types/registry"
+	"github.com/docker/docker/registry"
```
Comment (AkihiroSuda) [inline login.go:32]: Needs to update go.mod

File: login.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+			Destination: &options.password,
+		},
+		&cli.BoolFlag{
+			Name:  "password-stdin",
+			Usage: "Take the password from stdin",
+		},
+	},
+}
+
+//copied from github.com/cli/cli/command/registry/login.go
```
Comment (AkihiroSuda) [inline login.go:134]: No such repo

File: login.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+			return errors.New("--password and --password-stdin are mutually exclusive")
+		}
+	}
+
+	if clicontext.Bool("password-stdin") {
+		if clicontext.String("username") != "" {
+			return errors.New("Must provide --username with --password-stdin")
+		}
+
+		contents, err := ioutil.ReadAll(os.Stdin)
```
Comment (AkihiroSuda) [inline login.go:148]: clicontext.App.Reader

File: login.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func ConfigureAuthentification(authConfig *types.AuthConfig, options *loginOptions) error {
+	authConfig.Username = strings.TrimSpace(authConfig.Username)
+	if options.username = strings.TrimSpace(options.username); options.username == "" {
+		options.username = authConfig.Username
+	}
+
+	if options.username == "" {
+		return errors.Errorf("Error: Username is Required")
```
Comment (AkihiroSuda) [inline login.go:203]: Error string should start with a lower char

File: login.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+			Name:  "password-stdin",
+			Usage: "Take the password from stdin",
+		},
+	},
+}
+
+//copied from github.com/cli/cli/command/registry/login.go
+func verifyloginOptions(clicontext *cli.Context, options *loginOptions) error {
+	if clicontext.String("password") != "" {
+		fmt.Println("WARNING! Using --password via the CLI is insecure. Use --password-stdin.")
```
Comment (AkihiroSuda) [inline login.go:137]: logrus.Warn

File: login.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+		if err != nil {
+			return err
+		}
+
+		if err := dockerConfigFile.GetCredentialsStore(serverAddress).Store(clitypes.AuthConfig(*(authConfig))); err != nil {
+			return errors.Errorf("Error saving credentials: %v", err)
+		}
+
+		if response.Status != "" {
+			fmt.Println(response.Status)
```
Comment (AkihiroSuda) [inline login.go:109]: fmt.Fprintln(clicontext.App.Writer, response.Status)

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by AkihiroSuda** [comment]
Also, could you squash commits

**Comment by AkihiroSuda** [comment]
I tried this with Docker Hub, but this seems to always show "Login Succeeded"  even when no password was given.

**Comment by fahedouch** [comment]
> I tried this with Docker Hub, but this seems to always show "Login Succeeded" even when no password was given.

C'est fixed with prompt password. Tell me if is ok to squash my commits :)

**Comment by AkihiroSuda** [comment]
```
# nerdctl login -u ****
Password: ****
Login Succeeded

# nerdctl pull ****/****
FATA[0000] expected the hostname part of ac.ServerAddress ("index.docker.io") to be authConfigHostname="index.docker.io", got ""
```

File: go.sum [hunk 9]
```diff
@@ -565,12 +665,18 @@ github.com/json-iterator/go v1.1.10/go.mod h1:KdQUCv79m/52Kvf8AW2vK1V8akMuk1QjK/
 github.com/jstemmer/go-junit-report v0.0.0-20190106144839-af01ea7f8024/go.mod h1:6v2b51hI/fHJwM22ozAgKL4VKDeJcHhJFhtBdhmNjmU=
 github.com/jstemmer/go-junit-report v0.9.1/go.mod h1:Brl9GWCQeLvo8nXZwPNNblvFj/XSXhF0NWZEnDohbsk=
 github.com/jtolds/gls v4.20.0+incompatible/go.mod h1:QJZ7F/aHp+rZTRtaJ1ow/lLfFfVYBRgL+9YlvaHOwJU=
+github.com/juju/ansiterm v0.0.0-20180109212912-720a0952cc2a h1:FaWFmfWdAUKbSCtOU2QjDaorUexogfaMgbipgYATUMU=
```
Comment (AkihiroSuda) [inline go.sum:668]: LGPL v3 is incompatible with Apache License 2.0
https://github.com/juju/ansiterm/blob/master/LICENSE

Comment (AkihiroSuda) [reply go.sum:668]: opened https://github.com/manifoldco/promptui/issues/173

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Sorry, this seems to contain LGPL violation (not your fault though).

**Comment by fahedouch** [comment]
> ```
> # nerdctl login -u ****
> Password: ****
> Login Succeeded
> 
> # nerdctl pull ****/****
> FATA[0000] expected the hostname part of ac.ServerAddress ("index.docker.io") to be authConfigHostname="index.docker.io", got ""
> ```

@AkihiroSuda  fixed

**Comment by AkihiroSuda** [comment]
Thanks, I'll try later, but could you squash commits?

File: login.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+	if options.username = strings.TrimSpace(options.username); options.username == "" {
+		options.username = authConfig.Username
+	}
+
+	if options.username == "" {
+		return errors.Errorf("error: Username is Required")
+	}
+
+	if options.password == "" {
+		prompt := promptui.Prompt{
```
Comment (AkihiroSuda) [inline login.go:209]: This could be used without violating LGPL https://pkg.go.dev/golang.org/x/crypto@v0.0.0-20201221181555-eec23a3978ad/ssh/terminal

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: login.go (hunk 10 above)
Comment (AkihiroSuda) [reply login.go:209]: (Using any library is fine as long as it works without legal concern)

Comment (fahedouch) [reply login.go]: hi @AkihiroSuda  I switched to this lib.

File: go.mod [hunk 11]
```diff
@@ -13,17 +13,27 @@ require (
 	github.com/containernetworking/cni v0.8.0
 	github.com/containernetworking/plugins v0.9.0
 	github.com/docker/cli v20.10.0+incompatible
+	github.com/docker/docker v17.12.0-ce-rc1.0.20200730172259-9f28837c1d93+incompatible
 	github.com/docker/go-units v0.4.0
+	github.com/fvbommel/sortorder v1.0.2 // indirect
 	github.com/gogo/protobuf v1.3.1
+	github.com/golangci/golangci-lint v1.36.0 // indirect
+	github.com/manifoldco/promptui v0.8.0
```
Comment (AkihiroSuda) [inline go.mod:21]: This should no longer be needed to appear here

File: go.sum (hunk 9 above)
Comment (AkihiroSuda) [reply go.sum:663]: This should no longer be needed.
This will disappear if you run `rm -f go.sum && go mod tidy`

File: login.go [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+				return errors.Wrap(err, "error allocating terminal")
+			}
+			defer tty.Close()
+			fd = int(tty.Fd())
+		}
+		bytePassword, err := terminal.ReadPassword(fd)
+		if err != nil {
+			return errors.Wrap(err, "error reading password")
+		}
+		options.password = strings.TrimSpace(string(bytePassword))
```
Comment (AkihiroSuda) [inline login.go:228]: Why trim space?

**Comment by AkihiroSuda** [comment]
Thanks, but the LGPL lib still appears in go.mod and go.sum . Also please squash commits. (`git rebase -i HEAD~N` and mark commits with `s`. N is the number of your commits.)

**Review by AkihiroSuda — APPROVED** [review]
Thanks, I'm merging this, but next time please clean up the commit message as well. (`git commit -a -s --amend`)

File: logout.go [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/urfave/cli/v2"
+)
+
+var logoutCommand = &cli.Command{
+	Name:   "logout",
+	Usage:  "Log out from a Docker registry",
+	Action: logoutAction,
+}
+
+// code inspired from XXX
```
Comment (AkihiroSuda) [inline logout.go:32]: What was XXX?

**Pull request merged**